AWSAuthSDK/Sources/AWSMobileClient/AWSMobileClientExtensions.swift

```diff
@@ -288,36 +279,50 @@ extension AWSMobileClient {
-                        } else if let error = task.error {
-                            self.pendingAWSCredentialsCompletion?(nil, error)
-                        }
+                self.performFederatedSignInTasks(provider: providerName, token: token).continueWith() { task in
+                    // If any credentials operation is pending, we invoke the waiting block to resume with new credentials
+                    if let credentials = task.result {
+                        self.mobileClientStatusChanged(userState: .signedIn,
+                                                       additionalInfo: [self.ProviderKey:providerName, self.TokenKey: token])
+                        self.pendingAWSCredentialsCompletion?(credentials, nil)
+                        completionHandler(UserState.signedIn, nil)
+                    } else if let error = task.error {
+                        self.pendingAWSCredentialsCompletion?(nil, error)
+                        completionHandler(nil, error)
+                    }
+                    if self.pendingAWSCredentialsCompletion != nil {
                         self.credentialsFetchLock.leave()
                         self.pendingAWSCredentialsCompletion = nil
-                        return nil
-                    })
+                    }
+                    return nil
                 }
+                
             } else { // first time calling federatedSignIn
                 // If using developer authenticated identities, identityId is required.
                 // Check if identityId is specified by the developer, else return error.
                 if providerName == IdentityProvider.developer.rawValue {
                     if let devAuthenticatedIdentityId = federatedSignInOptions.cognitoIdentityId {
                         self.internalCredentialsProvider?.identityProvider.identityId = devAuthenticatedIdentityId
                     } else {
-                        error = AWSMobileClientError.invalidParameter(message: "For using developer authenticated identities, you need to specify the `CognitoIdentityId` in `FederatedSignInOptions`.")
+                        let error = AWSMobileClientError.invalidParameter(message: "For using developer authenticated identities, you need to specify the `CognitoIdentityId` in `FederatedSignInOptions`.")
+                        completionHandler(nil, error)
                         return
                     }
                 }
                 if let customRoleArn = federatedSignInOptions.customRoleARN {
                     self.customRoleArnInternal = customRoleArn
                     _AWSMobileClient.sharedInstance().setCustomRoleArnInternal(customRoleArn, for: self)
                 }
-                self.performFederatedSignInTasks(provider: providerName, token: token)
+                self.performFederatedSignInTasks(provider: providerName, token: token).continueWith() { task in
```

Comment:
nittiest of nits: no parens for trailing closure syntax :)

Reply:
Removed